store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java

```diff
@@ -2571,9 +2595,77 @@ public long getJoinTime() {
         }
     }
 
+    class BatchDispatchRequest {
+
+        private ByteBuffer byteBuffer;
+
+        private int position;
+
+        private int size;
+
+        private long id;
+
+        public BatchDispatchRequest(ByteBuffer byteBuffer, int position, int size, long id) {
+            this.byteBuffer = byteBuffer;
+            this.position = position;
+            this.size = size;
+            this.id = id;
+        }
+    }
+
+    class DispatchRequestOrderlyQueue {
+
+        DispatchRequest[][] buffer;
+
+        long ptr = 0;
+
+        AtomicLong maxPtr = new AtomicLong();
+
+        public DispatchRequestOrderlyQueue(int bufferNum) {
+            this.buffer = new DispatchRequest[bufferNum][];
+        }
+
+        public void put(long idx, DispatchRequest[] obj) {
+            while (ptr + this.buffer.length <= idx) {
+                synchronized (this) {
+                    try {
+                        this.wait();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+            int mod = (int) (idx % this.buffer.length);
+            this.buffer[mod] = obj;
+            maxPtr.incrementAndGet();
+        }
+
+        public DispatchRequest[] get(List<DispatchRequest[]> rets) {
+            synchronized (this) {
+                for (int i = 0; i < this.buffer.length; i++) {
+                    int mod = (int) (ptr % this.buffer.length);
+                    DispatchRequest[] ret = this.buffer[mod];
+                    if (ret == null) {
+                        this.notifyAll();
+                        return null;
+                    }
+                    rets.add(ret);
+                    this.buffer[mod] = null;
+                    ptr++;
+                }
+            }
+            return null;
+        }
+
+        public synchronized boolean isEmpty() {
+            return maxPtr.get() == ptr;
+        }
+
+    }
```

Comment:
Ditto. We should avoid abbreviated variable names like idx, obj or rets.

Reply:
And DispatchRequestOrderlyQueue#get always returns null, which looks weird. Changed to the following form to be more object-oriented.

```
public List<DispatchRequest[]> get()
```

Reply:
> And DispatchRequestOrderlyQueue#get always returns null, which looks weird. Changed to the following form to be more object-oriented.

How about I change it to void? Because it doesn't need a return value.

Reply:
> > And DispatchRequestOrderlyQueue#get always returns null, which looks weird. Changed to the following form to be more object-oriented.
> 
> How about I change it to void? Because it doesn't need a return value.

`public void get(List<DispatchRequest[]> rets)` seems like a function in C code style to avoid returning local variables causing dangling pointer problems. In Java, we are more accustomed to using the return value of the `get` method.

Reply:
I agree with you. The intention here is to reduce the new List<DispatchRequest[]>, so the return value is not set.